src/main/scala/rise/core/DSL/package.scala

```diff
@@ -80,8 +77,13 @@ package object DSL {
   }
 
   implicit class TypeAssertionHelper(t: Type) {
-    def !:[T <: Expr](e: ToBeTyped[T]): ToBeTyped[Expr] =
-      e >>= (e => toBeTyped(TypeAssertion(e, t)))
+    def !:[T <: Expr](e: ToBeTyped[T]): ToBeTyped[Expr] = {
+      val preserve = infer.getFTVs(t).map(_.name)
+      e >>= (e => {
+        val env = infer.collectFreeEnv(e)
+        infer(TypeAnnotation(e, t), env, preserve)
+      })
+    }
```

Comment:
I think that the `preserve` set should either be:
1. global to a rewritten expression:
```scala
val s = rewrite1 `;` rewrite2
val typed = start.toExpr
implicit val preserve = collectFreeEnv(typed)
s(typed) // uses 'preserve'
```
2. taken from the left-hand-side of each rewrite rule
```scala
def s: Strategy = lhs => infer(rhs :: lhs.t, collectFreeEnv(lhs), getFTVs(lhs))
```

Solution 2. requires less global code changes and is more similar to the solution I used in the equality saturation side:
https://github.com/rise-lang/shine/blob/91a3d2a620618b9514dfe8197c2c5291f82b1ede/src/main/scala/rise/eqsat/NamedRewrite.scala#L15-L23

In the future we can also consider having a common syntax for both equality saturation and Elevate rewrite rules:
```
def etaAbstraction = f : (a -> b) --> λx. (f x)
def etaReduction = λx. (f x) --> f
  when f ∌ x // x not in freeVars(f)

def mapFusion = map f (map g in) --> map λx. (f (g x)) in
def mapFission = {dt: data} -> map λx. (f (gx : dt)))) --> λin. map f (map λx. gx in)
  when f ∌ x
```
This would also enable type-checking Elevate rewrite rules ahead of match like I do for eqsat rewrite rules.

@michel-steuwer

Reply:
Pushed something along this lines, thanks!